streams/src/main/java/org/apache/kafka/streams/processor/internals/SourceNode.java

```diff
@@ -44,12 +46,20 @@ public SourceNode(String name, List<String> topics, Deserializer<K> keyDeseriali
         this(name, topics, null, keyDeserializer, valDeserializer);
     }
 
-    K deserializeKey(String topic, byte[] data) {
-        return keyDeserializer.deserialize(topic, data);
+    K deserializeKey(String topic, Headers headers, byte[] data) {
+        if (keyDeserializer instanceof ExtendedDeserializer) {
```

Comment:
To avoid this instanceof check on hot path, as with KafkaClient, you can change the 
    private Deserializer<K> keyDeserializer;
     private Deserializer<V> valDeserializer;

to Extended versions, and on construction wrap them, thus removing this logic on hot path.